src/NHibernate/Loader/Loader.cs

```diff
@@ -981,9 +981,18 @@ private void CheckVersion(int i, IEntityPersister persister, object id, object e
 						obj =
 							InstanceNotYetLoaded(rs, i, persister, key, lockModes[i], optionalObjectKey,
 												 optionalObject, hydratedObjects, session);
+
+						// IUniqueKeyLoadable.CacheByUniqueKeys caches all unique keys of the entity, regardless of
+						// associations loaded by the query. So if the entity is already loaded, it has forcibly already
+						// been cached too for all its unique keys, provided its persister implement it. With this new
+						// way of caching unique keys, it is no more needed to handle caching for alreadyLoaded path
+						// too.
+						(persister as IUniqueKeyLoadable).CacheByUniqueKeys(obj, session);
```

Comment:
Did you mean `(persister as IUniqueKeyLoadable)?.CacheByUniqueKeys(obj, session);`?

Reply:
Or `((IUniqueKeyLoadable) persister).CacheByUniqueKeys(obj, session);`. I have to check if it must always be castable or not.

Reply:
I think your version is the safer one. Maybe there is some loading cases, when there are no associations loaded or none with a unique key, where a custom persister could be something not implementing this interface, although all NHibernate `ILoadable` current implementations do also implement it.